server/sonar-process-monitor/src/test/java/org/sonar/application/SchedulerImplTest.java

```diff
@@ -0,0 +1,384 @@
+    assertThat(processLauncher.waitForProcess(WEB_SERVER).isAlive()).isFalse();
+
+    // verify that awaitTermination() does not block
+    underTest.awaitTermination();
+  }
+
+  @Test
+  public void restart_stops_all_if_new_settings_are_not_allowed() throws Exception {
+    Scheduler underTest = startAll();
+    doThrow(new IllegalStateException("reload error")).when(appReloader).reload(settings);
+
+    processLauncher.waitForProcess(WEB_SERVER).askedForRestart = true;
+
+    // waiting for all processes to be stopped
+    processLauncher.waitForProcessDown(ELASTICSEARCH);
+    processLauncher.waitForProcessDown(COMPUTE_ENGINE);
+    processLauncher.waitForProcessDown(WEB_SERVER);
+
+    // verify that awaitTermination() does not block
+    underTest.awaitTermination();
+  }
+
+  private SchedulerImpl newScheduler() {
+    return new SchedulerImpl(settings, appReloader, javaCommandFactory, processLauncher, appState)
+      .setProcessWatcherDelayMs(1L);
+  }
+
+  private Scheduler startAll() throws InterruptedException {
+    enableAllProcesses();
+    SchedulerImpl scheduler = newScheduler();
+    scheduler.schedule();
+    processLauncher.waitForProcess(ELASTICSEARCH).operational = true;
+    processLauncher.waitForProcess(WEB_SERVER).operational = true;
+    processLauncher.waitForProcess(COMPUTE_ENGINE).operational = true;
+    return scheduler;
+  }
+
+  private void waitForAppStateOperational(ProcessId id) throws InterruptedException {
+    while (true) {
+      if (appState.isOperational(id)) {
+        return;
+      }
+      Thread.sleep(1L);
+    }
+  }
+
+  private static class TestJavaCommandFactory implements JavaCommandFactory {
+    @Override
+    public JavaCommand createEsCommand() {
+      return ES_COMMAND;
+    }
+
+    @Override
+    public JavaCommand createWebCommand(boolean leader) {
+      return leader ? WEB_LEADER_COMMAND : WEB_FOLLOWER_COMMAND;
+    }
+
+    @Override
+    public JavaCommand createCeCommand() {
+      return CE_COMMAND;
+    }
+  }
+
+  private class TestJavaProcessLauncher implements JavaProcessLauncher {
+    private final EnumMap<ProcessId, TestProcess> processes = new EnumMap<>(ProcessId.class);
+    private final List<JavaCommand> commands = new ArrayList<>();
+    private ProcessId makeStartupFail = null;
+
+    @Override
+    public ProcessMonitor launch(JavaCommand javaCommand) {
+      commands.add(javaCommand);
+      if (makeStartupFail == javaCommand.getProcessId()) {
+        throw new IllegalStateException("cannot start " + javaCommand.getProcessId());
+      }
+      TestProcess process = new TestProcess(javaCommand.getProcessId());
+      processes.put(javaCommand.getProcessId(), process);
+      return process;
+    }
+
+    private TestProcess waitForProcess(ProcessId id) throws InterruptedException {
+      while (true) {
+        TestProcess p = processes.get(id);
+        if (p != null) {
+          return p;
+        }
+        Thread.sleep(1L);
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Remove this use of "Thread.sleep()". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2925)